File: algorithm/104_maximum_depth_of_binary_tree.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	q := []*TreeNode{root}
	depth := 1
	start, end := 0, 1
	for {
		child := 0
		for i := 0; start < end; i++ {
			node := q[i]
			if node.Left != nil {
				q = append(q, node.Left)
				child++
			}
			if node.Right != nil {
				q = append(q, node.Right)
				child++
			}
		}
		if child == 0 {
			break
		}
		depth++
		start = end
		end += child
	}
	return depth
}

File: algorithm/104_maximum_depth_of_binary_tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if root is None:
            return 0
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))


leftNode = TreeNode(9)
rightNode = TreeNode(20, TreeNode(15), TreeNode(7))
node = TreeNode(3, left=leftNode, right=rightNode)
soln = Solution()
print(soln.maxDepth(node))
